### DISCUSSION TITLE
Which layer to choose - guideline

### DISCUSSION THREAD

**Discussion by OneManShow0815** [mrdbourke/pytorch-deep-learning#1038] [Q&A] [answered]
How do I choose a layer when for machine learning. Or, for what purpose is which layer best suitable?

**Comment by LuluW8071** [comment]
@OneManShow0815  Can you give more context

**Reply by OneManShow0815** [reply]
Hi LuluW0871, there a re many different Linear models in PyTorch. Every one has a different algorithm in it. Is there a description for which case files which Linear model suits the best?
Or after which guidlines do you choose a Layer when writing a software using PyTorch? Danial used ..nLinear in his first part, later on other ones.
As of now, I do not have specific task to write, just to understand how to choose one of the different layers

**Reply by LuluW8071** [reply]
There is no strict rule regarding the number of layers you should use in a model. Generally, **linear models** are utilized as **feature selectors** in the middle or final layers of **CNNs** and **RNNs** or in **combinations of both**. Additionally, **sequential layers** are used to stack linear or CNN layers for convenience, during the implementation of the `forward pass` function.

Since u r at the beginning of the course, understanding these concepts might be hard at first, but they will become clearer as u progress through the **CNN (computer vision)** section. I recommend reviewing the theory on artificial neural networks (ANNs) to grasp how linear networks function before diving into coding. Daniel has explained about ANNs as well before coding it.

**Comment by OneManShow0815** [comment]
Hi LuluW8071, Thanks for the quick response.
My question was more which layer y use, like  Dropout layer or recurrent layer or other.

Regards
Sven

**Reply by LuluW8071** [reply, marked as answer]
**Linear Layer** for basic model or mid/final feature selector
**Dropout layer** for making our model unbiased towards training data only introducing non linearity
**BatchNorm** for stable (recenter and rescale with normlization) and fastertraining
**CNN** for image data(picture, audio)
**RNNs** for sequential(mostly textual and time-series) data
**Transformer** for text, images, videos, audios 

Watch some vids and read research papers to get the concepts behind it

**Reply by OneManShow0815** [reply]
Thank you LuluW8071,

Great answer, 100% clear. Great.

Regards
Sven
